Readme.md
# littleBIGtable

#### littleBIGtable is a small (~5k gzipped) javascript table built using AlpineJS

![Screenshot](screenshot.gif)

See the [Complete example](http://iangrindley.uk/littleBIGtable/examples/complete.php) in action, this uses Bulma for the styling as shown above.

*littleBIGtable is very new and while it works well for my purposes you may find issues, please report them in the issue tracker.* 

littleBIGtable is CSS agnostic and can be used with any css framework, the default examples use Bulma, there is also a [Bootstrap example](http://iangrindley.uk/littleBIGtable/examples/bootstrap.php) using Bootstrap 4 and a [Basic example](http://iangrindley.uk/littleBIGtable/examples/basic.php) with no css. I imagine it would be great with TailwindCSS too.

As you have complete control over the HTML you can structure and style it however you wish.

## Goals

 * Allow you the developer to completely control the look and feel of the table by wrapping a standard HTML table in an AlpineJS component.
 * Make the viewing and searching of large remote datasets easy, it does not use local data, it only fetches data from your chosen endpoint.

## Why
I created this as too many table plugins render the table in an opinionated fashion, I wanted to be able to write the table in HTML and 'sprinkle' the javascript over it to make it interactive.

## Changelog

## [0.1.0] - 2021-02-12

First dependable release, I have started using this in Production.

### Added

- New setting [headers](#adding-http-headers) to allow sending custom HTTP headers with the requests.
- New setting [args](#changing-the-url-query-arguments) to allow setting the url query argument names to suit your application.

### Changed

- The default storage key prefix is now capitalised to match the new project name, it was 'lbt' now it is 'lBt', hopefully this makes it more obvious as to what it relates to.

### Fixed

- An issue with the pager display when the total number of results was less than number of results to show 

## Installation

### Using the CDN

Link to the main javascript file

```html
<script src="https://unpkg.com/littlebigtable@0.1.1/dist/littleBIGtable.min.js" defer></script>
```

If you want to use the supplied icons you must refer to them when passing in your options to the littleBIGtable component:

```js
let opts = {
  icons: "https://unpkg.com/littlebigtable@0.1.1/dist/icons.svg"
}
```
```html
<div x-data="littleBIGtable(opts)" x-init="init()">

</div>
```

### Manually
Copy the `icons.svg` and `littelBIGtable.min.js` files from the dist folder into your project somewhere and add the following script element.

```html
<script src="/path/to/your/folder/littleBIGtable.min.js" defer></script>
```

You will need to change the icons.svg locations as well, see the examples for details.

```html
<svg><use xlink:href="/path/to/your/install/folder/icons.svg#[name-of-icon]"></use></svg>
```

### Using npm

There is an npm package for littleBIGtable available for install, however it is not recommended until the package is updated.

```sh
npm install littlebigtable
```


## Documentation

### Get started

This is the starting point for all littleBIGtables, you must wrap your table in an AlpineJS component, provide a url to your api endpoint and specify your table columns and rows. The code below is an example of the minium required.

```html
<div x-data="littleBIGtable({url:'/location/of/your/api'})" x-init="init()">
	<table>
		<thead>
			<tr>
				<th>Title</th>
				<th>Year</th>
				<th>Rating</th>
			</tr>	
		</thead>
		<tbody>
			<template x-for="row in rows">
				<tr>
					<td x-text="row.title"></td>
					<td x-text="row.year"></td>
					<td x-text="row.rating"></td>
				</tr>	
			</template>
		</tbody>
	</table>
</div>
<script src="/path/to/alpine.min.js" defer></script>
<script src="/path/to/littleBIGtable.min.js"></script>
```

Check the source of the [Basic example](http://iangrindley.uk/littleBIGtable/examples/basic.php) to see this in action.


### Settings

The following settings can be overriden by passing in an object in the `x-data` attribute:

## Fetching data

littleBIGtable expects a JSON response from your endpoint, the object must contain two properties, an Integer `total` and an Array `data` with total being the total number of rows in the remote dataset and data containing the subset of rows to be displayed.

In PHP you might do something like this:

```php
<?php
// the total number of records in the dataset
$total = 100;
// an array of rows to display
$data = [
	[
		'title' => '...',
		'year' => '...',
		'rating' => '...'
	],
	// etc...
];

echo json_encode([
	'total' => $total,
	'data' => $data
]);
``` 

In the examples folder you'll find a simple PHP JSON server, it can be started on your local machine by using the built in PHP server, type the following into your terminal:

```sh
php -S localhost:8080
```

## Formatting cell contents

Before the cell data is rendered it can be modified using formatters, they are functions that return a string and can be used to add links, classes, show images or anything you can think of.

NB: If you are returning HTML you must us the `x-html` directive on the `<td>` element, if you use `x-text` the HTML will be escaped.

### Creating a formatter

To create a formatter add your function into the formatters setting passed to littleBIGtable. Your function will receive the cell value and the complete row data. Whatever you return will be displayed in the cell.

Assuming you have columns called `year` and `rating` you can create formatters as follows:

```js

let settings = {
	url: '/path/to/your/endpoint',
	formatters: {
		year: function(year, row) {
			if (year < 2000) {
				return `<span class="blue">${year}</span>`;
			}
			return year;
		},
		rating: function(rating, row) {
			if (rating > 4) {
				return `<span class="green">${rating}</span>`;
			}
			if (rating < 3) {
				return `<span class="red">${rating}</span>`;
			}
			return rating;
		}
	}
}


``` 

That example would add the class 'blue' to any year in the last century and add the classes 'red' or 'green' to the rating depending on its value.


## Adding functionality

#### Displaying the statusbar

Show a statusbar to show handy messages to the user, loading state or errors and provide a handy reference to where they are in the results.

Simply add the following somewhere inside your component `<div>`.

`<p x-html="meta.status"></p>`

The statusbar messages can be changed in the settings. There are two types of *summary* text, one displays where the user is in terms of rows visible and total, the other is more book like and shows the current page and total pages.

`Showing rows 1 to 10 of 95 results` or `Showing page 2 of 4`

To choose between the two set the value of `messages.summary` to either `rows` or `pages` when passing your settings.

You can also customise the `loading` and `failed` state messages by passing `messages.loading` and `messages.failed` in your settings. 


#### Adding a row limiter 

Row limiters limit the number of rows displayed at any one time, the default is 25 but any number between 1 and 100 can be chosen. Any value chosen by the user is saved in localStorage and reused. 

Simply add a drop down and add the '' function. This is limited to between 1 and 100 rows.

Check the [Complete example](http://iangrindley.uk/littleBIGtable/examples/complete.php) for more details.

#### Adding a search box

A search box can be used to filter your remote data.

Add the code for the search box, ensuring it is inside the  component div (the one with the x-data attribute) and add the following ''

Check the [Complete example](http://iangrindley.uk/littleBIGtable/examples/complete.php) for more details.

#### Making columns sortable

Columns can be sorted by adding a `<button>` element to toggle the sort direction. 

Add the `<button>` element to your `th` using the `x-html` and `@click` directives ([docs](https://github.com/alpinejs/alpine/#learn)). 

```html
<th>
	Year
	<button 
		type="button" 
		x-html="getSortIcon('year')"
		@click="doSort('year')">
	</button>
</th>
<!-- The above code assumes the year column name is 'year' -->
```

The `getSortIcon()` method will automatically show the correct icon state using an icon from icons.svg.

The `doSort()` method sets the URL sort parameter and fetches the sorted data. The url parameter appears as `&sort=column_name:asc` or  `&sort=column_name:dsc` depending on the requested sort direction.

Check the [Complete example](http://iangrindley.uk/littleBIGtable/examples/complete.php) for more details.

## Adding HTTP headers

If you need to add or adjust the default HTTP headers pass the headers key in the options, the default headers are shown below:
```html
<script>
  let options = {
  headers: {
    'Content-Type': 'application/json',
    'X-Requested-With': 'littleBIGtable'
  },
}
</script>
<div x-data="littleBIGtable(options)" x-init="init()">
  ...
</div>
``` 

## Changing the URL query arguments

If you need to change the default query argument names pass the args key in the options
```html
<script>
let options = {
  args: {
    offset: 'page',
    limit: 'show_per_page',
    search: 'query',
    sort: 'order_by'
  },
}
</script>
<div x-data="littleBIGtable(options)" x-init="init()">
  ...
</div>
``` 

### Custom events

It is possible to create custom event handlers, for example to handle a row 'click' add the `@click` handler to your table rows and provide the name of your custom function which can be included anywhere on the page or in a separate file. 

```html
<tr @click="doSomething($event)" data-id="123">
```

By passing the `$event` 'magic property' ([docs](https://github.com/alpinejs/alpine/#event)) you'll be able to see all the event propeties, for example to get the `data-id` attribute of the above clicked row 

```js
function doSomething(event) {
	console.log('The row was clicked, below is the event and the data-id of the row');
	console.log(event);
	console.log(event.target.parentNode.attributes['data-id'].value);
}

```    


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.0] - 2021-02-12

First dependable release, I have started using this in Production.

### Added

- New setting [headers](#adding-http-headers) to allow sending custom HTTP headers with the requests.
- New setting [args](#changing-the-url-query-arguments) to allow setting the url query argument names to suit your application.

### Changed

- The default storage key prefix is now capitalised to match the new project name, it was 'lbt' now it is 'lBt', hopefully this makes it more obvious as to what it relates to.

### Fixed

- An issue with the pager display when the total number of results was less than number of results to show 


examples/complete.php
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.9.1/css/bulma.min.css">
  <title>'littleBIGtable' Complete example</title>
  <meta name="description" content="littleBIGtable is a small (~4k gzipped) javascript table using AlpineJS" />
  <!-- <meta http-equiv="Content-Security-Policy" content="base-uri: 'none'; object-src: 'none'; script-src 'self' 'unsafe-eval' 'unsafe-inline' 'nonce-abc123'"> -->
</head>
<body>
    <section class="section">
        <div class="container">
            <h1 class="title is-3">littleBIGtable</h1>
            <p class="content is-size-5">This is the complete example with custom settings</p>
            <p class="content">Passing in custom options is easy, just create an options object with the properties you want to override and pass it through the <code>x-data</code> attribute</p>
            <pre class="mb-6"><code>let options = {
    // set this to wherever your data comes from
    url: 'http://localhost:8080/examples/json.php', 
    // this is the prefix for the page limit preference, you can may want to change this on a per table basis
    key_prefix: 'lbt',
    // change the default status bar messages as you see fit
    messages: {
        loading: 'Loading...',
        failed: 'Loading failed',
        summary: 'rows'     // this may be null, 'rows' or 'pages'
    },
    // formatters change the way cell data is displayed
    formatters: {},
    // the location of the SVG icons file
    icon: '../dist/icons.svg',
}</code></pre>
            <p class="content">
                Now pass in your options to the component &lt;div&gt;
            </p>
            <pre class="mb-6"><code>&lt;div x-data="littleBIGtable(options)" x-init="init()"&gt;</code></pre>
            <div x-data="littleBIGtable(options)" x-init="init()">
                <div class="level">
                    <div class="level-left">
                        <div class="field is-horizontal">
                            <div class="field-label is-normal">
                                <label class="label">Search</label>
                            </div>
                            <div class="field-body">
                                <div class="field is-narrow">
                                    <div class="control">
                                        <input class="input" type="text" placeholder="Start typing..." x-model="params.search" @keyup.debounce.350="doSearch()">
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div class="level-right">
                        <div class="field is-horizontal">
                            <div class="field-label is-normal">
                                <label class="label">Show</label>
                            </div>
                            <div class="field-body">
                                <div class="field is-narrow">
                                    <div class="control">
                                        <div class="select is-fullwidth">
                                            <select @change="setLimit()" x-model="params.limit">
                                                <option value="10">10 per page</option>
                                                <option value="15">15 per page</option>
                                                <option value="25">25 per page</option>
                                                <option value="50">50 per page</option>
                                                <option value="100">100 per page</option>
                                            </select>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                <table class="table is-fullwidth is-hoverable">
                    <thead>
                        <tr>
                            <th class="has-text-left">State <button class="button is-small is-text" type="button" x-html="getSortIcon('state')" @click="doSort('state')"></button></th>
                            <th class="has-text-left">County <button class="button is-small is-text" type="button" x-html="getSortIcon('county')" @click="doSort('county')"></button></th>
                            <th class="has-text-centered">Year <button class="button is-small is-text" type="button" x-html="getSortIcon('year')" @click="doSort('year')"></button></th>
                            <th class="has-text-centered">Capacity <button class="button is-small is-text" type="button" x-html="getSortIcon('turbine_capacity')" @click="doSort('turbine_capacity')"></button></th>
                            <th class="has-text-centered">Turbines <button class="button is-small is-text" type="button" x-html="getSortIcon('project_capacity')" @click="doSort('project_capacity')"></button></th>
                        </tr>
                    </thead>
                    <tbody>
                        <template x-for="row in rows">
                            <tr @click="doRowClicked($event)" data-id="xyz">
                                <td x-html="row.state"></td>
                                <td x-text="row.county"></td>
                                <td class="has-text-centered" x-text="row.year"></td>
                                <td class="has-text-centered" x-html="row.turbine_capacity"></td>
                                <td class="has-text-centered" x-text="row.project_capacity"></td>
                            </tr>
                        </template>
                    </tbody>
                </table>
                <div class="level">
                    <div class="level-left">
                        <p x-html="meta.status"></p>
                    </div>
                    <div class="level-right">
                        <nav class="pagination" role="navigation" aria-label="pagination">
                            <ul class="pagination-list">
                                <li>
                                    <a class="pagination-link" aria-label="Goto first page" @click="goFirstPage()" :disabled="getCurrentPage() == 1">
                                        <svg class="icon"><use xlink:href="../dist/icons.svg#page-first"></use></svg>
                                    </a>
                                </li>
                                <li>
                                    <a class="pagination-previous" @click="goPrevPage()" :disabled="getCurrentPage() == 1">
                                        <svg class="icon"><use xlink:href="../dist/icons.svg#page-prev"></use></svg>
                                    </a>
                                </li>
                                <li>
                                    
                                </li>
                                <li>
                                    <a class="pagination-next" @click="goNextPage()" :disabled="getCurrentPage() == getTotalPages()">
                                        <svg class="icon"><use xlink:href="../dist/icons.svg#page-next"></use></svg>
                                    </a>
                                </li>
                                <li>
                                    <a class="pagination-link" aria-label="Goto last page" @click="goLastPage()" :disabled="getCurrentPage() == getTotalPages()">
                                        <svg class="icon"><use xlink:href="../dist/icons.svg#page-last"></use></svg>
                                    </a>
                                </li>
                            </ul>
                        </nav>
                    </div>
                </div>
            </div>
        </div>
    </section>
    <script src="../dist/alpine.min.js" defer></script>
    <script src="../dist/littleBIGtable.min.js"></script>
    <script nonce="abc123">
        options = {
            'url': 'http://localhost:8080/examples/json.php',
            'limit': 25,
            'formatters': {
                'state': function(value, row) {
                    return '<strong>' + value + '</strong>';
                },
                'turbine_capacity': function(value, row) {
                    if (parseInt(value) < 1500) {
                        return '<span class="has-text-warning has-text-weight-medium">' + value  + "</span>";
                    }
                    if (parseInt(value) > 2000) {
                        return '<span class="has-text-success has-text-weight-medium">' + value  + "</span>";
                    }
                    return '<span class="has-text-link has-text-weight-medium">' + value  + "</span>";
                }
            }
        }
        // this is an example of a row click event, in this example the row is specified as follows:
        // <tr @click="doRowClicked($event)" data-id="xyz">
        function doRowClicked(e) {
            console.log('The row was clicked, below is the event and the data-id of the row');
            console.log(e);
            console.log(e.target.parentNode.attributes['data-id'].value);
        }
    </script>
</body>
</html>
